### ISSUE TITLE
Regression related type objects and no strict optional

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#19751]
When type checking these files, there's a false positive in `x.py` when using a cold cache (`rm -rf .mypy_cache` first). This is a regression from mypy 1.17 and is a release blocker. The behavior is inconsistent between cached and uncached runs.

`typ.py`:
```py
from typing import NamedTuple

NT = NamedTuple("NT", [("x", str | None)])
```

`mod.py`:
```py
# mypy: no-strict-optional

from typ import NT

def f() -> NT:
    return NT(x='')
```

`x.py`:
```py
import mod

# Argument "x" to "NT" has incompatible type "None"; expected "str"
a = mod.NT(x=None)
```
It looks like the `NT` parameter type gets simplified to just `str` from `str | None` when the type is used in a module with strict optional checking disabled.

This bisects to #19514 (cc @ilevkivskyi). Maybe the caching should be disabled in modules that don't use strict optional checking?

**Comment by ilevkivskyi** [comment]
Yeah, cache invalidation is one of the three hardest problems. I guess we can still read the cached constructor in modules with strict optional disabled, but we should not _write_ to cache in such modules.
